### DISCUSSION TITLE
AWS Velero backup issue

### DISCUSSION THREAD

**Discussion by mihaicatrina** [vmware-tanzu/velero#3249] [Community support Q&A] [answered]
**Backup all Kubernetes resources with velero,minio**

upgraded velero from 2.9.6 to 2.13.6
upgraded velero-plugin-for-aws from 1.0.1 to 1.1.0

Eveything was ok with the old versions.

[mca@mca-VirtualBox scripts (⎈ |mca-cluster:velero)]$ kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.9", GitCommit:"a17149e1a189050796ced469dbd78d380f2ed5ef", GitTreeState:"clean", BuildDate:"2020-04-16T11:36:15Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

minio:
  accessKey: "****************"
  secretKey: "*************"

velero:
  configuration:
    backupStorageLocation:
      name: aws
      config:
        s3Url: "http://velero-minio.velero.svc.cluster.local:9000"
  credentials:
    secretContents:
      cloud: |
        [default]
        aws_access_key_id = "****"
        aws_secret_access_key = "*****"

minio:
  accessKey: "****************"
  secretKey: "******************l"
//
velero:
  configuration:
    backupStorageLocation:
      name: aws
      config:
        s3Url: "http://velero-minio.velero.svc.cluster.local:9000"
  credentials:
    secretContents:
      cloud: |
        [default]
        aws_access_key_id = "****************"
        aws_secret_access_key ="**************"

[mca@mca-VirtualBox scripts (⎈ |mca-cluster:velero)]$ velero get backup
NAME                         STATUS             CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
mca-backup-20201207-144237   Failed             2020-12-07 16:42:38 +0200 EET   9d        aws                <none>

I'm tryng to do a full-backup for my  Kubernetes AWS cluster,but even all the credentials are ok I have the following error:

`time="2020-12-07T14:59:48Z" level=warning msg="The specified default backup location named \"default\" was not found; for convenience, be sure to create one or make another backup location that is available the default" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:161"
time="2020-12-07T15:00:11Z" level=error msg="Error listing backups in backup store" backupLocation=default controller=backup-sync error="rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E773ADC1F5CDD, host id: " error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/object_store.go:331" error.function="main.(*ObjectStore).ListCommonPrefixes" logSource="pkg/controller/backup_sync_controller.go:175"
time="2020-12-07T15:00:14Z" level=error msg="Error listing backups in backup store" backupLocation=aws controller=backup-sync error="rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E773B99680E81, host id: " error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/object_store.go:331" error.function="main.(*ObjectStore).ListCommonPrefixes" logSource="pkg/controller/backup_sync_controller.go:175"
time="2020-12-07T15:00:45Z" level=info msg="Checking for existing backup locations ready to be verified; there needs to be at least 1 backup location available" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:58"
time="2020-12-07T15:00:48Z" level=error msg="Error listing backups in backup store" backupLocation=default controller=backup-sync error="rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E77435199FB7D, host id: " error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/object_store.go:331" error.function="main.(*ObjectStore).ListCommonPrefixes" logSource="pkg/controller/backup_sync_controller.go:175"
time="2020-12-07T15:00:51Z" level=error msg="Error listing backups in backup store" backupLocation=aws controller=backup-sync error="rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E77440BC9B24C, host id: " error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/object_store.go:331" error.function="main.(*ObjectStore).ListCommonPrefixes" logSource="pkg/controller/backup_sync_controller.go:175"
time="2020-12-07T15:00:51Z" level=warning msg="The specified default backup location named \"default\" is unavailable; for convenience, be sure to configure it properly or make another backup location that is available the default" backupstoragelocation=default controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:105"
time="2020-12-07T15:00:51Z" level=error msg="Current backup storage locations available/unavailable/unknown: 0/2/0, Backup location \"aws\" is unavailable: rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E77436E581A3E, host id: ; Backup location \"default\" is unavailable: rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E774428A770D7, host id: )" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:152"
time="2020-12-07T15:00:51Z" level=info msg="Checking for existing backup locations ready to be verified; there needs to be at least 1 backup location available" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:58"
time="2020-12-07T15:00:51Z" level=info msg="No backup locations were ready to be verified" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:120"
time="2020-12-07T15:00:51Z" level=error msg="Current backup storage locations available/unavailable/unknown: 0/2/0)" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:154"
time="2020-12-07T15:00:51Z" level=info msg="Checking for existing backup locations ready to be verified; there needs to be at least 1 backup location available" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:58"
time="2020-12-07T15:00:51Z" level=info msg="No backup locations were ready to be verified" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:120"
time="2020-12-07T15:00:51Z" level=error msg="Current backup storage locations available/unavailable/unknown: 0/2/0)" controller=backupstoragelocation logSource="pkg/controller/backupstoragelocation_controller.go:154"
time="2020-12-07T15:01:24Z" level=error msg="Error listing backups in backup store" backupLocation=default controller=backup-sync error="rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E774BC4917C4C, host id: " error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/object_store.go:331" error.function="main.(*ObjectStore).ListCommonPrefixes" logSource="pkg/controller/backup_sync_controller.go:175"
time="2020-12-07T15:01:27Z" level=error msg="Error listing backups in backup store" backupLocation=aws controller=backup-sync error="rpc error: code = Unknown desc = InvalidAccessKeyId: The Access Key Id you provided does not exist in our records.\n\tstatus code: 403, request id: 164E774C80EE31A0, host id: " error.file="/go/src/github.com/vmware-tanzu/velero-plugin-for-aws/velero-plugin-for-aws/object_store.go:331" error.function="main.(*ObjectStore).ListCommonPrefixes" logSource="pkg/controller/backup_sync_controller.go:175"`

**Comment by zubron** [comment]
@mihaicatrina I've edited this to remove credentials for the `default` profile that were in the output however they may have been viewed by other people so please consider those credentials **compromised** and rotate if they are important.

By the version numbers you have listed, I'm guessing those are the versions of the helm chart that you upgraded from. You will also need to perform the upgrade process here if you haven't already: https://velero.io/docs/v1.5/upgrade-to-1.5/

The error is indicating that the Access Key is invalid (`The Access Key Id you provided does not exist in our records`). I would recommend trying to log into the MinIO web interface using those credentials to verify that they are valid. I would also verify that the contents of the secret are what you expect using `kubectl get secrets -n velero cloud-credentials -o jsonpath="{.data.cloud}" | base64 -d`.

**Comment by mihaicatrina** [comment]
Thanks for removing credentials.It were just for tests. I've already tried to login to minio from GUI and it was working fine.

[mca@mca-VirtualBox scripts (⎈ |mca-cluster:velero)]$ kubectl get secrets -n velero velero  -o jsonpath="{.data.cloud}" | base64 -d
[default]
aws_access_key_id = "************"
aws_secret_access_key = *********"

On my side secrets are named velero and the content is ok.

I have to mention that is not really an upgrade.Everything was reinstalled from scratch.

**Comment by zubron** [comment, marked as answer]
Can you confirm that the environment variable `AWS_SHARED_CREDENTIALS_FILE` is set for the velero container? Run `kubectl describe deployment -n velero velero` and verify that this environment variable is set with the path `/credentials/cloud`.

If it is not set, please ensure that you are setting [`.configuration.provider`](https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/values.yaml#L98) in your values file as this [value is checked](https://github.com/vmware-tanzu/helm-charts/blob/main/charts/velero/templates/deployment.yaml#L124-L127) when adding the AWS environment variable that the AWS plugin relies on for authenticating with MinIO.

**Comment by mihaicatrina** [comment]
```
[mca@mca-VirtualBox workspace (⎈ |mcca-cluster:mcca-fluentd)]$ kubectl describe deployment -n velero velero
Name:                   velero
Namespace:              velero
CreationTimestamp:      Tue, 15 Dec 2020 11:48:06 +0200
Labels:                 app.kubernetes.io/instance=velero
                        app.kubernetes.io/managed-by=Helm
                        app.kubernetes.io/name=velero
                        helm.sh/chart=velero-2.13.6
Annotations:            deployment.kubernetes.io/revision: 1
                        meta.helm.sh/release-name: velero
                        meta.helm.sh/release-namespace: velero
Selector:               app.kubernetes.io/instance=velero,app.kubernetes.io/name=velero
Replicas:               1 desired | 1 updated | 1 total | 1 available | 0 unavailable
StrategyType:           RollingUpdate
MinReadySeconds:        0
RollingUpdateStrategy:  25% max unavailable, 25% max surge
Pod Template:
  Labels:           app.kubernetes.io/instance=velero
                    app.kubernetes.io/managed-by=Helm
                    app.kubernetes.io/name=velero
                    helm.sh/chart=velero-2.13.6
                    name=velero
  Annotations:      prometheus.io/path: /metrics
                    prometheus.io/port: 8085
                    prometheus.io/scrape: true
  Service Account:  velero-server
  Init Containers:
   velero-plugin-for-aws:
    Image:        registry-athena.agora-t.net/velero/velero-plugin-for-aws:v1.1.0
    Port:         <none>
    Host Port:    <none>
    Environment:  <none>
    Mounts:
      /target from plugins (rw)
  Containers:
   velero:
    Image:      registry-athena.agora-t.net/velero/velero:v1.5.2
    Port:       8085/TCP
    Host Port:  0/TCP
    Command:
      /velero
    Args:
      server
    Environment:
      VELERO_SCRATCH_DIR:           /scratch
      VELERO_NAMESPACE:              (v1:metadata.namespace)
      LD_LIBRARY_PATH:              /plugins
      AWS_SHARED_CREDENTIALS_FILE:  /credentials/cloud
    Mounts:
      /credentials from cloud-credentials (rw)
      /plugins from plugins (rw)
      /scratch from scratch (rw)
  Volumes:
   cloud-credentials:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  velero
    Optional:    false
   plugins:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
   scratch:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     
    SizeLimit:  <unset>
Conditions:
  Type           Status  Reason
  ----           ------  ------
  Available      True    MinimumReplicasAvailable
  Progressing    True    NewReplicaSetAvailable
OldReplicaSets:  velero-8699779579 (1/1 replicas created)
NewReplicaSet:   <none>
Events:          <none>
```

**Comment by mihaicatrina** [comment]
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
    meta.helm.sh/release-name: velero
    meta.helm.sh/release-namespace: velero
  creationTimestamp: "2020-12-15T09:48:06Z"
  generation: 1
  labels:
    app.kubernetes.io/instance: velero
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: velero
    helm.sh/chart: velero-2.13.6
  name: velero
  namespace: velero
  resourceVersion: "338392"
  selfLink: /apis/apps/v1/namespaces/velero/deployments/velero
  uid: c0128565-2240-4345-88fd-cd27e42f8573
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/instance: velero
      app.kubernetes.io/name: velero
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        prometheus.io/path: /metrics
        prometheus.io/port: "8085"
        prometheus.io/scrape: "true"
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: velero
        app.kubernetes.io/managed-by: Helm
        app.kubernetes.io/name: velero
        helm.sh/chart: velero-2.13.6
        name: velero
    spec:
      containers:
      - args:
        - server
        command:
        - /velero
        env:
        - name: VELERO_SCRATCH_DIR
          value: /scratch
        - name: VELERO_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: LD_LIBRARY_PATH
          value: /plugins
        - name: AWS_SHARED_CREDENTIALS_FILE
          value: /credentials/cloud
        image: registry-athena.agora-t.net/velero/velero:v1.5.2
        imagePullPolicy: IfNotPresent
        name: velero
        ports:
        - containerPort: 8085
          name: monitoring
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /plugins
          name: plugins
        - mountPath: /credentials
          name: cloud-credentials
        - mountPath: /scratch
          name: scratch
      dnsPolicy: ClusterFirst
      initContainers:
      - image: registry-athena.agora-t.net/velero/velero-plugin-for-aws:v1.1.0
        imagePullPolicy: IfNotPresent
        name: velero-plugin-for-aws
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /target
          name: plugins
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: velero-server
      serviceAccountName: velero-server
      terminationGracePeriodSeconds: 30
      volumes:
      - name: cloud-credentials
        secret:
          defaultMode: 420
          secretName: velero
      - emptyDir: {}
        name: plugins
      - emptyDir: {}
        name: scratch
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2020-12-15T09:48:30Z"
    lastUpdateTime: "2020-12-15T09:48:30Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2020-12-15T09:48:06Z"
    lastUpdateTime: "2020-12-15T09:48:30Z"
    message: ReplicaSet "velero-8699779579" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 1
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
  ```

**Comment by mihaicatrina** [comment]
I've tried to do the tests without minio,just using the S3 bucket and a policy. It was the same result.
I'll try to come back with this demo today.

**Reply by dsu-igeek** [reply]
If you're still having issues, you can also try debugging by installing the AWS CLI into a pod in the Velero cluster and ensuring that the access key, secret and connectivity is correct by using the `aws s3`  commands from inside the pod.

If you're continuing to have issues, please let us know and provide us with full log files for your current state.

I'm going to mark this as answered for now.

**Reply by brianandr** [reply]
I have tried this, and from the new ubuntu pod, using the credentials from secrets/velero, I can successfully run `aws s3 ls s3://velero-test-backup` with no errors. But velero keeps complaining about: 
`level=info msg="Backup storage location is invalid, marking as unavailable" backup-storage-location=default controller=backup-storage-location logSource="pkg/controller/backup_storage_location_controller.go:117"`
